doc/SONiC_Port_Mirroring_HLD.md

```diff
@@ -0,0 +1,319 @@
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+
+# Revision
+| Rev |     Date    |       Author       | Change Description                         |
+|:---:|:-----------:|:------------------:|--------------------------------------------|
+| 0.1 | 05/17/2019  |   Rupesh Kumar      | Initial version                            |
+
+
+# About this Manual
+This document provides general information about extending mirroring implementation in SONiC.
+# Scope
+This document describes the high level design of Mirroring Enhancements feature. 
+
+
+# Definition/Abbreviation
+### Table 1: Abbreviations
+| **Term**                 | **Meaning**                         |
+|--------------------------|-------------------------------------|
+|   SPAN                   |  Switched Port ANalyzer                |
+|   ERSPAN                 |  Encapsulated Remote Switched Port ANalyzer                |
+
+
+# 1 Requirement Overview
+## 1.1 Functional Requirements
+
+1. Port/Port-channel mirroring support
+     - Add support to mirror ingress traffic on port/port-channel to SPAN/ERPSAN mirror session.
+     - Add support to mirror egress traffic on port/port-channel to SPAN/ERSPAN mirror session.
+     - Add support to mirror both ingress/egress traffic on port/port-channel to SPAN/ERSPAN mirror session.
+
+2. Dynamic session management
+    - Allow multiple source to single destination.
+    - Each session supports mirroring from single port to single destination port.
+    - Session-id created in SAI per destination port will be used when the same destination port is configured in other session.
+      This effectively utilizes the hardware resource to be shared across multiple sessions.
+    - Mirror session on source portchannel will be active if at least one port is part of portchannel.
+    - Mirror session on source portchannel will become inactive when portchannel has no members.
+    - ERSPAN session will be active/inactive based on destination IP reachability.
+
+3. ACL rules can continue to use port/ERSPAN sessions as the action.
+
+4. Configuration CLI for mirror session
+    - CLI allows all flavors of mirror sessions.
+    - CLI validation for all mandatory parameters in ERSPAN configuration.
+    - CLI validation for all mandatory parameters in port/portchannel mirroring.
+    - CLI to allow mirror session configuration only with destination port.
+
+
+## 1.2 Configuration and Management Requirements
+- Existing CLI 'config mirror_session add/remove'to be extended to include source port/portchannel.
+- Existing CLI 'config mirror_session add/remove' to be extended to include destination port/portchannel.
+- Existing CLI 'show mirror session' is extended to support all flavors of mirror sessions.
+
+
+## 1.3 Scalability Requirements
+- Up to max ASIC capable mirror sessions to be supported.
+- Once max mirror sessions are created and user attempts to create new session, error will be logged in syslog.
```

Comment:
it is not discussed in the design document how this is implemented. can you add a section to discuss how you are going to implement this?

Reply:
When the user exceeds max mirror sessions then SAI returns SAI_STATUS_INSUFFICIENT_RESOURCES, This is similar behaviour as what is currently present in ERSPAN also. The error is logged to syslog from SyncD. OrchAgent treats the error as fatal, similar to the existing ERPSAN code.